### PR TITLE
docs(public-api): Ensure public API functions/methods have complete docstrings

### PR DESCRIPTION
**Pull request by tonyyli-wandb**
Description
-----------
- Fixes [WB-18977](https://wandb.atlassian.net/browse/WB-18977)

Changes:
- Updates existing `pydocstyle` checks (prev context in, e.g. #5036) to require docstrings on "public API" functions and methods, specifically those defined in `wandb.apis.public.api`
- Adds or updates type hints in `api.py` and related modules: not comprehensively, but at least anywhere it seemed appropriate (e.g. on functions where docstrings were added or edited). 

Caveats:
- This PR only enforces new requirements on the `wandb.apis.public.api` module to maintain a reasonable scope of changes for this PR.  Some adjacent modules (i.e. `wandb.apis.public.*`) were also edited, but were not the deliberate focus of this PR.
- Missing docstrings are still allowed in `@property` descriptors -- also to maintain a reasonable scope for the PR.  Ideally this exemption could be dropped in the future.
- Missing docstrings are still allowed in `wandb.apis.public.api.RetryingClient` -- even though this is in a "public" module, it doesn't look like a class that users are encouraged to instantiate/manage directly.

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
N/A.  Only type annotations and docstrings were changed.  No changes to existing functionality or underlying implementation.

(Note: admittedly new type annotations can introduce import errors -- e.g. via circular imports -- but this should be caught by existing SDK tests.)

[WB-18977]: https://wandb.atlassian.net/browse/WB-18977?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dmitryduev — CHANGES_REQUESTED** [review]
Thanks @tonyyli-wandb. Could you please make sure the CI is green and re-request a review?

File: pyproject.toml [hunk 1]
```diff
@@ -171,8 +171,13 @@ select = ["F", "E", "W", "B", "I", "N", "D", "C90", "UP"]
 ignore = [
     "B904",
     "E501",
-    "D1",    # Allow missing docstrings.
-    "D417",  # Don't require descriptions for all arguments.
```
Comment (moredatarequired) [inline pyproject.toml:175]: We're removing D417 here? That doesn't seem feasible

File: wandb/apis/public/api.py [hunk 2]
```diff
@@ -283,14 +283,33 @@ def __init__(
...
-        """Create a new run."""
-        if kwargs.get("entity") is None:
-            kwargs["entity"] = self.default_entity
-        return public.Run.create(self, **kwargs)
+    def create_run(
+        self,
+        *,
+        run_id: Optional[str] = None,
+        project: Optional[str] = None,
+        entity: Optional[str] = None,
```
Comment (moredatarequired) [inline wandb/apis/public/api.py:300]: 👍

File: wandb/apis/public/api.py [hunk 3]
```diff
@@ -954,7 +1007,18 @@ def artifact_versions(self, type_name, name, per_page=50):
...
+        self, type_name: str, name: str, per_page: Optional[int] = 50
+    ) -> "public.Artifacts":
+        """Return an `Artifacts` collection from the given parameters.
+
+        Arguments:
+            type_name: (str) The type of artifacts to fetch.
+            name: (str) An artifact collection name. May be prefixed with entity/project.
+            per_page: (int, optional) Sets the page size for query pagination.  None will use the default size.
+                Usually there is no reason to change this.
+
```
Comment (moredatarequired) [inline wandb/apis/public/api.py:1020]:
```suggestion
```

**Review by moredatarequired — APPROVED** [review]
This looks great, much appreciated!

I think there might have been some historical reason for (intentionally) not adding type hints in this file, but I'm not sure what it could break for users, so unless someone from SDK team remembers I'm all for it.

File: pyproject.toml (hunk 1 above)
Comment (tonyyli-wandb) [reply pyproject.toml:175]: I thought so too, but evidently this is only enforced if there's a proper `Arguments/Args/Parameters` section in the docstring.

So I think this rule doesn't really demand "function args need to be fully documented," it just demands "function args need to be fully documented or not at all...no half measures."

File: wandb/apis/public/api.py [hunk 4]
```diff
@@ -301,6 +320,7 @@ def create_report(
         width: Optional[str] = "readable",
         blocks: Optional["wandb.apis.reports.util.Block"] = None,
     ) -> "wandb.apis.reports.Report":
+        """Create a new report."""
```
Comment (kptkin) [inline wandb/apis/public/api.py:323]: why only partial if already touching this?

File: wandb/apis/public/api.py [hunk 5]
```diff
@@ -669,7 +689,8 @@ def projects(self, entity=None, per_page=200):
             )
         return self._projects[entity]
 
-    def project(self, name, entity=None):
+    def project(self, name, entity=None) -> "public.Project":
+        """Return the `Project` with the given name (and entity, if given)."""
```
Comment (kptkin) [inline wandb/apis/public/api.py:693]: same as above

File: wandb/apis/public/api.py [hunk 6]
```diff
@@ -717,10 +738,15 @@ def create_team(self, team, admin_username=None):
         """
         return public.Team.create(self, team, admin_username)
 
-    def team(self, team):
+    def team(self, team: str) -> "public.Team":
+        """Return the matching `Team` with the given name.
+
+        Arguments:
+            team: (str) The name of the team.
```
Comment (kptkin) [inline wandb/apis/public/api.py:745]: why not add return?

File: wandb/apis/public/api.py [hunk 7]
```diff
@@ -909,29 +937,54 @@ def sweep(self, path=""):
...
     @normalize_exceptions
-    def artifact_type(self, type_name, project=None):
+    def artifact_type(
+        self, type_name: str, project: Optional[str] = None
+    ) -> "public.ArtifactType":
+        """Return the matching `ArtifactType`.
+
+        Arguments:
+            type_name: (str) The name of the artifact type to retrieve.
+            project: (str, optional) If given, a project name or path to filter on.
```
Comment (kptkin) [inline wandb/apis/public/api.py:957]: why no retruns?

File: wandb/apis/public/api.py [hunk 8]
```diff
@@ -909,29 +937,54 @@ def sweep(self, path=""):
...
+        self, project_name: str, type_name: str, per_page: Optional[int] = 50
+    ) -> "public.ArtifactCollections":
+        """Return a collection of matching artifact collections.
+
+        Arguments:
+            project_name: (str) The name of the project to filter on.
+            type_name: (str) The name of the artifact type to filter on.
+            per_page: (int, optional) Sets the page size for query pagination.  None will use the default size.
+                Usually there is no reason to change this.
+        """
```
Comment (kptkin) [inline wandb/apis/public/api.py:973]: same re: returns

**Review by kptkin — APPROVED** [review]
thanks for adding this, this is really great!
would be great to do a complete work on the things you touched!

File: wandb/apis/public/api.py (hunk 4 above)
Comment (tonyyli-wandb) [reply wandb/apis/public/api.py:323]: Mainly just an oversight -- thanks for catching this (and the others)!

**Comment by tonyyli-wandb** [comment]
@dmitryduev polite bump -- when you get a chance, could you take another look to see if requested changes have been addressed? (or if other changes would be good needed)

**Review by dmitryduev — APPROVED** [review]
🙏

**Pull request merged**